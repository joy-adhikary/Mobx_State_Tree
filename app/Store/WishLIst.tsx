import {
    applySnapshot,
    onSnapshot,
    getSnapshot,
    flow,
    types,
} from "mobx-state-tree";

// ! Jkn Mobx State Tree er maddhome kono state change korbo tkn clone use korbo mobx er , then oi clone er getsnapshot niye applysnapshot er madhome update kore dibo 
//! clone basically current state er akta copy banai dei 
// ! Clone simply takes the type of object you give it and takes its snapshot, and then instantiates a new instance of the same type with the same snapshot, so you get basically a literal copy of the same thing.

interface WishlistItemType {
    name: string;
    id: number;
    price: number;
    img: string;
}

export const WishListItem = types.model("WishListItem",
    {
        name: types.string,
        id: types.integer,
        price: types.number,
        img: types.optional(types.string, ""),
    }
)
    .actions((self) => (
        {
            ChangeName(newName: string) {
                self.name = newName;
            },

            ChangePrice(NewPrice: number) {
                self.price = NewPrice;
            },

            ChangeImage(newImage: string) {
                self.img = newImage;
            },
        }
    ))

export const WishList = types.model("WishList",
    {
        items: types.optional(types.array(WishListItem), []),
    }
)
    .actions((self) => (
        {
            AddItem(CommingItem: WishlistItemType) {
                self.items.push(CommingItem);
            },

            RemoveItem(RemoveItemById: number) {
                self.items = self.items.filter(item => item.id !== RemoveItemById)
            },
        }
    ))

    .views(self => (
        {
            get TotalPrice() {
                return self.items.reduce((total: number, item: WishlistItemType) => total + item.price, 0);
            }
        }
    ))
