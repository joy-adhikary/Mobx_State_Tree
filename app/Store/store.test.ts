import { describe, expect, test } from '@jest/globals';
import { getSnapshot, onSnapshot, onPatch } from "mobx-state-tree"
import { WishList, WishListItem } from "./WishLIst"

describe('Testing WishListItem Model', () => {
    test("should create a new wish list item", () => {
        const Item = WishListItem.create(
            {
                id: 1,
                name: "joy",
                price: 28.73,
            }
        );

        expect(Item.price).toBe(28.73);
        expect(Item.img).toBe("");

        Item.ChangeName("Joy kumar Adhikary");
        expect(Item.name).toBe("Joy kumar Adhikary");

        Item.ChangePrice(10);
        expect(Item.price).toBe(10);
    });
});


describe('Testing WishListItem Model ', () => {
    test("should create a new wishList ", () => {
        
        const Item = WishList.create(
            {
                items: [
                    {
                        id: 1,
                        name: "joy",
                        price: 10,
                    }
                ]
            }
        );

        expect(Item.items.length).toBe(1)
        expect(Item.items[0].id).toBe(1)

        Item.AddItem(
            {
                id: 11,
                name: "joya",
                price: 2000.0,
                img: "joy",
            }
        )
        expect(Item.items.length).toBe(2)

        expect(getSnapshot(Item)).toEqual(
            {
                items: [
                    {
                        id: 1,
                        name: "joy",
                        price: 10,
                        img: "",
                    },
                    {
                        id: 11,
                        name: "joya",
                        price: 2000.0,
                        img: "joy",
                    }
                ]
            }
        )

        // Upper one and this one basically both are the same 
        expect(getSnapshot(Item)).toMatchSnapshot()

        // Total Price Views 
        expect(Item.TotalPrice).toBe(2010.0)
        
        Item.RemoveItem(11)
        expect(Item.items.length).toBe(1)

    });
});

