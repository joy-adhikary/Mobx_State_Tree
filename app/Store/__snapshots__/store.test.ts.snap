// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing WishListItem Model  should create a new wishList  1`] = `
{
  "items": [
    {
      "id": 1,
      "img": "",
      "name": "joy",
      "price": 10,
    },
    {
      "id": 11,
      "img": "joy",
      "name": "joya",
      "price": 2000,
    },
  ],
}
`;
